Comments: Persist after status change

Copons opened the pull request:
Introduce a `persistedComments` list that keeps track of all the comments toggling between `approved` and `unapproved` statuses (including if the toggle happens as a side-effect of `like` or `reply`).
Persisted comments are left in their original list until navigating to a different tab.

![jul-20-2017 18-54-57](https://user-images.githubusercontent.com/2070010/28431675-0e3bc7c0-6d7d-11e7-8c14-a4a0c09ee556.gif)

cc @Automattic/lannister

gwwar commented:
Took this out for a spin, and behavior feels pretty good. Can you take a look @drw158 ?

Do we need undo notices if items remain in the list? Folks can instead perform the inverse action if they made a mistake.

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -93,7 +103,11 @@ export class CommentList extends Component {
+
+	removeFromPersistedComments = commentId => this.setState( {
+		persistedComments: reject( this.state.persistedComments, { ID: commentId } ),
+	} );
```
I just recently became aware of the function-passing syntax here and we should probably use it because of the way React can batch state updates. that is, because we're accessing the previous state we should pass in a function…

```js
this.setState( ( { persistedComments } ) => ( {
	persistedComments: reject( persistedComments, { ID: commentId } ),
} ) );
```

dmsnell commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -82,16 +82,31 @@ export class CommentDetail extends Component {
 	toggleApprove = () => {
 		const { commentId, commentStatus, postId, setCommentStatus } = this.props;
-		setCommentStatus( commentId, postId, 'approved' === commentStatus ? 'unapproved' : 'approved' );
+		const isPersisted = 'approved' === commentStatus || 'unapproved' === commentStatus;
```
maybe `shouldPersist` could more closely capture our intention here. it hasn't _yet_ persisted, after all, has it?

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -132,18 +146,30 @@ export class CommentList extends Component {
 
-	setCommentStatus = ( commentId, postId, status, options = { showNotice: true } ) => {
+	setCommentStatus = ( commentId, postId, status, options = { isUndo: false, persist: false, showNotice: true } ) => {
+		const { isUndo, persist, showNotice } = options;
```
correspondingly to above, I often find that `doPersist` is a good name for an argument whose boolean intention is to mark something or not

dmsnell commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -82,16 +82,31 @@ export class CommentDetail extends Component {
-		setCommentStatus( commentId, postId, 'approved' === commentStatus ? 'unapproved' : 'approved' );
+		const isPersisted = 'approved' === commentStatus || 'unapproved' === commentStatus;
+
+		setCommentStatus( commentId, postId, 'approved' === commentStatus ? 'unapproved' : 'approved', {
```
just a nitpick, but we usually got `siteId, postId, commentId` in that order. you can chose on your own

Copons replied:
This is a leftover from the Faker days, where we only used the `commentId`.
Real actions though also require `siteId` and `postId`.
`siteId` is passed inside `mapDispatchToProps` for maximum convenience, therefore we're left with `postId` and `commentId`.

I know that our usual argument order is "from larger to smaller context" (`siteId, postId, commentId`), but in this case using `postId` before `commentId` feels a bit weird to me.

Though, should we want to modify this, I'd do it in another PR to avoid cluttering this one with unrelated changes.

dmsnell replied:
it would have been nice if we had included `siteId` and `postId` in the faker so that we wouldn't have had to make any changes here now. if we will need them anyway, why isn't `siteId` here?

Copons replied:
> why isn't siteId here?

It's passed directly in `mapDispatchToProps`.

> we can uniquely identify a single comment by siteId and commentId only

But the actions still require `postId`.

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -55,12 +55,16 @@ export class CommentList extends Component {
 
 	state = {
 		isBulkEdit: false,
+		persistedComments: [],
```
I like using a list of comments here as the persistent storage :+1:

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -93,7 +103,11 @@ export class CommentList extends Component {
 	isCommentSelected = commentId => !! this.state.selectedComments[ commentId ];
 
-	isSelectedAll = () => this.props.comments.length === size( this.state.selectedComments );
+	isSelectedAll = () => this.getComments().length === size( this.state.selectedComments );
```
is `selectedComments` not an array? it's a little odd seeing two different ways of computing collection size right next to each other.

Copons replied:
Another leftover from the Faker days.
The Faker passed an ID-keyed object of comments instead of an array, and so I just reused the structure for `selectedComments`.

But because `persistedComments` must be merged with `props.comments` (which is an array), it just is cleaner to do all arrays.

I fully expected a comment about this weirdness, but again it's a change for another PR. 🙂

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -132,18 +146,30 @@ export class CommentList extends Component {
+				...comment,
+				i_like: 'approved' !== status ? false : comment.i_like,
+				status,
+			}, isUndo );
```
let's be careful here. we're actually storing a copy of data that's available in the Redux state tree _and_ we're changing properties on it. we're opening ourselves up to synchronization bugs (for instance, if someone updates the comment in another session and Calypso becomes aware of it, this will still show stale data).

granted this is not a huge risk but I think it's probably also easy to store the list of comment ids that are persisted. in that case (and hopefully if we move to passing lists of ids anyway) the views will always be up to date because they will get their data from state directly and no data will be cloned in independent silos.

Copons replied:
Great suggestion! Gonna look into it asap

Copons replied:
Ok this is the problem here: it's true that the data is available in the Redux state, but not in the component.
The `getSiteComments` selector takes care of filtering the comments list by status, so that `CommentList` only knows about the comments belonging to the currently displayed list.
In other words: as soon as I change a comment status, that comment disappears from the portion of state passed to `CommentList`.

In order to persist it (and make it work seamlessly), I both have to store it _and_ manually update its `status` and `i_like` values (according to the status change action I performed).
This is unavoidable for the same reason as above: I can't copy the comment _after_ the status change because it wouldn't be there anymore.

One way to change this would be that `getSiteComments` returns all comments (or, eventually, comment IDs), and delegates to `CommentList.render()` the burden of filtering by status.
This way we'd have access to everything, but at the same time we'd make `CommentList` aware of (a potentially huge amount of) stuff it doesn't (and might never) really care.

Copons replied:
> if someone updates the comment in another session and Calypso becomes aware of it

This would be an issue anyway though.
Say you're casually browsing the Pending list. Somebody else approves a bunch of comments and Calypso picks those changes up: suddenly you'd be seeing comments disappearing from the list for no apparent reason.

In the persistence case, as you said, there's not a huge risk. At least from a "potential screw up" point of view.
Let me try an example:

- You've just approved a comment:
`realComment.status = "unapproved" → "approved"`
- The comment is then persisting in your "Pending" list, with a fixed value of:
`persistedComment.status = "approved"`
- Somebody else decides to unapprove it:
`realComment.status = "approved" → "unapproved"`
- You, unaware of the fact that the comment has in fact been already unapproved again, decide to undo your approval:
`persistedComment.status = "approved" → "unapproved"`
but the real action would go like this instead:
`realComment.status = "unapproved" → "unapproved"`
No harm done.

Of corse if several people keep modifying the same comment back and forth at some point everything will go out of sync, but that's to be expected anyway. 😄

dmsnell replied:
why can't we just store the ids of the comments that have been approved/unapproved and manually grab them from state in addition to the list of comments the view wants?

dmsnell commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -198,32 +224,54 @@ export class CommentList extends Component {
+				} );
+			}
 		}
 	}
```
great googly moogly @copons I found the collection of `}`s I have been missing!

do you think we could explore un-nested early-abort measures instead of deeply-nesting here? check out the complexity we're creating when we branch for each condition…

Copons replied:
> I found the collection of `}`s I have been missing!

You're welcome!

Copons commented:
@dmsnell following your suggestion I've tried to convert everything to a list of comment IDs, starting from ac99705.

In **`CommentList`** we pull `getSiteComments( state, siteId, status )` and, using only the IDs, we build the `CommentList.props.comments` array.
In **`CommentDetail`** we select the single comment with the  `getSiteComment( state, siteId, commentId )` and proceed as before.

When we persist a comment, we just add its ID into `CommentList.state.persistedComments` that, merged with `props.comments` will result in the list of `CommentDetail`, with the freshest possible values of `status` and `i_like`.

---

This is by itself mostly working, with the most notable exception of the Undo.
Before, `CommentList` had the onus of knowing to what values revert a comment. Now though, it doesn't know anything more than the IDs.
Therefore, it's `CommentDetail` that needs to tell the list about it.

My solution is to create a stripped-down comment object, containing only the values we actually need for the various actions, and we send that back to `CommentList`.
```es6
CommentDetail.getCommentObject = () => ( {
	commentId: this.props.commentId,
	isLiked: this.props.commentIsLiked,
	postId: this.props.postId,
	status: this.props.commentStatus,
} );
```
_(yes: it could use a better name)_

`commentId` and `postId` are required by the various Redux actions and API calls.
`isLiked` and `status` are used for the Undo.

Thanks to this, we can reduce the number of parameters for the various actions, and we'd also fix the wrong order of `commentId, postId` between component methods and Redux actions.

---

I've added a ton of TODOs on everything related to the bulk actions because it used stuff I've now removed.
Bulk actions are slated to m2, and this PR is already way too big on its own.

drw158 commented:
Everything felt pretty natural. Even if the undo button is removed I think the notice is helpful for confirmation that the action happened. I think it's nice to have a quick option to undo.

One thing I thought is that maybe if you are approving a ton of comments at once, the persistence could be annoying. We might consider not persisting on bulk actions, I'm not sure. We will see if the users have any feedback on that.

Also, I guess paging won't affect the persistence?

Copons commented:
I feel this is in a good state right now, and I'm fairly happy with the new stripped-down comment object containing only the bare minimum of info being passed back and forth.

I know it got a bit bloated, but most changes are minor (e.g. added plenty of missing `_.get()` default values and plenty of TODO comments).
Should it feel too cumbersome to review, I could try dividing it up in two PRs:
- One that replaces the old logic with the stripped down comment object.
- One that actually takes care of the persistence.

> Also, I guess paging won't affect the persistence?

@drw158 We'll know more when we wire up the new comment tree endpoint, but it shouldn't right out of the box.
We receive the list of all comment IDs for a given status. When we change a comment status, its ID is removed from that list, but at the same time we add it to a temporary persisted comments list.
Then we show all the comments contained in _both_ lists; actually: a new list resulted from the merge of the two list). This new merged list doesn't change its length after a (persistent) action is performed.
The pagination will be done on the merged list, and as long as the length is the same, the pagination will be the same too.

dmsnell commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -80,18 +81,40 @@ export class CommentDetail extends Component {
+		const { commentStatus, setCommentStatus } = this.props;
+		const shouldPersist = 'approved' === commentStatus || 'unapproved' === commentStatus;
+
+		setCommentStatus( this.getCommentObject(), 'approved' === commentStatus ? 'unapproved' : 'approved', {
```
just a quick note that this can seem ambiguous here. ternary precedence varies by language so we could anticipate this code doing one of two things:

  - either return `unapproved` or `approved` based on whether `commentStatus` === approved
  - return a boolean if `approved` === (either `unapproved` or `approved` based on whether `commentStatus` is truthy)

a pair of parens would remove all ambiguity

gwwar replied:
I would also, suggest moving the next status on it's own line, like we do with shouldPersist. I was getting distracted with the extra arg after the ternary.

dmsnell commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -80,18 +81,40 @@ export class CommentDetail extends Component {
+		isLiked: this.props.commentIsLiked,
+		postId: this.props.postId,
+		status: this.props.commentStatus,
+	} );
```
it would be much better here I think to pass in the props at the call site and let this be a normal function outside of the React component…

```js
const getCommentObject = ( { commentId, commentIsLiked, commentStatus, postId } ) => ( {
	commentId,
	commentStatus,
	isLiked: commentIsLiked,
	postId,
} )
```

gwwar replied:
What sort of comment object is this? I think keeping it here or moving it out as a normal function works, but this is likely a smell that we could be doing this translation in the data-layer or in a selector.

Copons replied:
This awfully named `commentObject` is simply an object that contains only the comment info that we need when changing a comment status (and toggling its like, as a side effect such as like+approve).
`commentId` and `postId` are required by the Redux actions.
`status` and `isLiked` are the comment "state" right before the actions are performed, and are used to undo the actions.

This is basically a way to keep the function declarations readable.
E.g.
`setStatus( commentObject, newStatus )`
instead of
`setStatus( postId, commentId, newStatus, previousStatus, previousIsLiked )`

It might seem unnecessary, but we end up using those 4 values pretty much everywhere (set status, toggle like, show notice).

As it's something that we need for communication between `CommentDetail` and `CommentList`, in my opinion stay out of the data layer.

Another approach could be to just pass the entire `comment` back and forth, but it seems a overkill.
Also, since we don't have the entire `comment` object available inside the component, we'd need to re-nest all `CommentDetail` props.
For example, from:
```es6
mapStateToProps = ( ... ) => ( {
	commentContent: get( comment, 'content' ),
	commentDate: get( comment, 'date' ),
	commentId: get( comment, 'ID' ),
	commentIsLiked: get( comment, 'i_like' ),
	commentStatus: get( comment, 'status' ),
	commentUrl: get( comment, 'URL' ),
} );
```
to
```es6
mapStateToProps = ( ... ) => ( {
	comment,
} );
```

(we may or may not want to do this, but I'd argue not here!)

gwwar replied:
Ah got it, this feels pretty similar to a redux action object (it's just missing the type), so we might even call this `getCommentStatusAction` where we might modify the fields to have both the last status and next status.

gwwar replied:
Anyway use your best judgement, if I see a getCommentObject out of context, I might assume this is a full comment representation.

gwwar commented on client/blocks/comment-detail/comment-detail-reply.jsx:
```diff
@@ -35,9 +35,13 @@ export class CommentDetailReply extends Component {
-		const { authorDisplayName, commentStatus, translate } = this.props;
+		const {
+			authorDisplayName,
+			comment: { status },
```
I think this is ok, but let's try and keep the destructure to one level.

gwwar commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -17,6 +17,7 @@ import CommentDetailPost from './comment-detail-post';
 import CommentDetailReply from './comment-detail-reply';
 import { decodeEntities, stripHTML } from 'lib/formatting';
 import { getPostCommentsTree } from 'state/comments/selectors';
+import getSiteComment from 'state/selectors/get-site-comment';
```
Oh curious, were there any bundle size benefits of this? Normally I see the following, which makes it easy to import multiple selectors.
```jsx
import { getSiteComment } from 'state/selectors';
```

Copons replied:
I'm not sure about the bundle size benefits.
Just done this because it's the only selector we pull from `state/selectors`.

gwwar commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -249,23 +264,23 @@ const mapStateToProps = ( state, ownProps ) => {
 		postTitle,
-		repliedToComment: comment.replied, // TODO: not available in the current data structure
-		siteId: comment.siteId || siteId,
+		repliedToComment: get( comment, 'replied' ), // TODO: not available in the current data structure
```
Is this still not available?

gwwar commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -55,12 +55,17 @@ export class CommentList extends Component {
 	state = {
 		isBulkEdit: false,
+		persistedComments: [],
+		// TODO: replace {} with [] after persistedComments is merged
```
Is this referring to another PR, or are we splitting this work up for smaller diffs?

Copons replied:
Basically the change from "list of comments" to "list of comment IDs" in this PR breaks the current selection implementation.
It needs to be slightly adapted and, on top of the additional code to review, it would have needed more testing efforts, since the selection is currently disabled via feature flag.

gwwar commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -91,13 +99,25 @@ export class CommentList extends Component {
-	isSelectedAll = () => this.props.comments.length === size( this.state.selectedComments );
+	isSelectedAll = () => this.getComments().length === size( this.state.selectedComments );
+
+	removeFromPersistedComments = commentId => this.setState( ( { persistedComments } ) => ( {
```
I'm afraid I'm not following this, could you explain what's happening from `this.setState( ( { persistedComments } )`?

gwwar replied:
I would suggest adding a comment or js doc here, or expanding the function a bit, since we tend not to curry too often.

gwwar replied:
Oh I missed a paren, no wonder I was confused. We're passing a function into setState, I misread earlier that there was an extra level of currying with a statement.

gwwar replied:
How about the following for spacing?

```jsx
	removeFromPersistedComments = commentId => this.setState(
		( { persistedComments } ) => ( {
			persistedComments: persistedComments.filter( c => c !== commentId ),
		} )
	);
```
